syntax = "proto3";

package posts;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

service PostsService {
  rpc CreatePost (CreatePostRequest) returns (Post);
  rpc UpdatePost (UpdatePostRequest) returns (Post);
  rpc DeletePost (DeletePostRequest) returns (google.protobuf.Empty);
  rpc GetPost (GetPostRequest) returns (Post);
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}

message Post {
  int64 id = 1;
  string title = 2;
  string description = 3;
  int64 creator_id = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int64 creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message UpdatePostRequest {
  int64 id = 1;
  int64 creator_id = 2; // для проверки авторизации
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message DeletePostRequest {
  int64 id = 1;
  int64 creator_id = 2; // для проверки авторизации
}

message GetPostRequest {
  int64 id = 1;
  int64 viewer_id = 2; // для проверки доступа к приватным постам
}

message ListPostsRequest {
  int64 creator_id = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total = 2;
  int32 page = 3;
  int32 total_pages = 4;
}
